app/controllers/lti_launch_controller.rb

```diff
@@ -0,0 +1,224 @@
+    if @jwt[:body]['https://purl.imsglobal.org/spec/lti/claim/version'] != "1.3.0"
+      raise LtiError.new("launch claim version is not 1.3.0", :bad_request)
+    end
+    return unless @jwt[:body]["https://purl.imsglobal.org/spec/lti/claim/roles"].nil?
+
+    raise LtiError.new("Roles claim not found", :bad_request)
+  end
+
+  # make sure that we are given the context_memberships_url
+  # otherwise, we can't call / access NRPS
+  def validate_nrps_access
+    # rubocop:disable Layout/LineLength
+    if @jwt[:body]['https://purl.imsglobal.org/spec/lti-nrps/claim/namesroleservice'].nil? ||
+       @jwt[:body]['https://purl.imsglobal.org/spec/lti-nrps/claim/namesroleservice']['context_memberships_url'].nil? ||
+       @jwt[:body]['https://purl.imsglobal.org/spec/lti-nrps/claim/namesroleservice']['context_memberships_url'].empty?
+      raise LtiError.new("NRPS context membership url not found", :bad_request)
+    end
+    # rubocop:enable Layout/LineLength
+  end
+
+  def validate_jwt_signature(id_token)
+    rsa_public = OpenSSL::PKey::RSA.new(Rails.configuration.lti_settings["platform_public_key"])
+    begin
+      JWT.decode id_token, rsa_public, true, { algorithm: 'RS256' }
+    rescue JWT::ExpiredSignature
+      # Handle expired token, e.g. logout user or deny access
+      raise LtiError.new("JWT signature expired", :bad_request)
+    end
+  rescue JWT::ImmatureSignature
+    # Handle invalid token, e.g. logout user or deny access
+    raise LtiError.new("JWT signature invalid", :bad_request)
+  end
+
+  # final LTI launch flow endpoint
+  # validate id_token, jwt, check we have NRPS access
+  # redirect to users/:id/lti_launch_initialize for final linking
+  def launch
+    # Code based on:
+    # https://github.com/IMSGlobal/lti-1-3-php-library/blob/master/src/lti/LTI_Message_Launch.php
+    validate_state(params)
+    id_token = params["id_token"]
+    validate_jwt_format(id_token)
+    validate_jwt_signature(id_token)
+    validate_nonce
+    validate_registration
+    validate_link_request
+    validate_nrps_access
+    if !current_user.present?
+      raise LtiError.new("Not logged in!", :bad_request)
+    end
+
+    @user = current_user
+    redirect_to controller: "users", action: "lti_launch_initialize",
+                launch_context: @jwt[:body], id: @user.id
+  end
+
+  # LTI launch entrypoint to initiate open id connect login
+  # build our authentication response and redirect back to
+  # platform
+  def oidc_login
+    # code based on: https://github.com/IMSGlobal/lti-1-3-php-library/blob/master/src/lti/LTI_OIDC_Login.php
+    # validate OIDC
+    validate_oidc_login(params)
+
+    # Build OIDC Auth Response
+    # Generate State.
+    # Set cookie (short lived)
+    state = SecureRandom.uuid
+    stateCookie = "lti1p3_#{state}"
+    cookies[stateCookie] = { value: state, expires_in: 1.hour }
+
+    # generate nonce, store in cache
+    nonce = "nonce-#{SecureRandom.uuid}"
+    Rails.cache.write('nonce', nonce)
```

Comment:
Might want to add an expiry to the cache write (see #1634 -- although I have yet to investigate whether that works)

Also, should we use a unique key for each user? It seems that each generated nonce overwrites the previous nonce

Reply:
I'll test adding an expiry, we could use a unique key for each user though. The point of the nonce is to prevent replay attacks, so the generated nonce should overwrite the previous one, but I guess if you get multiple user launches a previous nonce of one user could get overwritten by another user.

Reply:
Yup, my concern is with multiple users launching at the same time